File: day05.hs
import Data.List

wordsWhen :: (Char -> Bool) -> String -> [String]
wordsWhen p s =  case dropWhile p s of
                  "" -> []
                  s' -> w : wordsWhen p s''
                            where (w, s'') = break p s'

myParse :: [String] -> [[Integer]]
myParse [] = []
myParse (x:xs) = [a .. b] : myParse xs
                            where a = read (head (wordsWhen (=='-') x)) :: Integer
                                  b = read (last (wordsWhen (=='-') x)) :: Integer

myFindIndices :: Char -> String -> Int
myFindIndices x s = head (elemIndices x s)

parseSetup :: String -> Int -> String
parseSetup s i = 
  if x `elem` ['A'..'Z'] then [x]
  else ""
    where x = s!!i


main = do
        contents <- readFile "input.txt"
        let rows = wordsWhen (=='\n') contents
        let setup = filter ( not . isInfixOf "move") rows
        let ins = filter (isInfixOf "move") rows
        let lastSetup = last setup
        let initSetup = init setup
        let numberOfStacks = read [(last (init lastSetup))] :: Integer
        let stacks = map show [1..numberOfStacks]
       
        let stackIndex = [myFindIndices (head x) lastSetup | x <- stacks]
        print(stackIndex)
        print(initSetup)
        let base = ["" | x <- [1..numberOfStacks]]
        print(base)

File: day05.py
with open("input.txt") as f:
    input = f.read().splitlines()


moves = [x for x in input if "move" in x]
setup = [x for x in input if "[" in x]

setup = [*zip(*setup)]

init = []
for s in setup:
    filtered = "".join(filter(str.isalpha, s))
    if filtered:
        init.append(filtered[::-1])

for move in moves:
    q = int(move.split()[1])
    f = int(move.split()[3])
    t = int(move.split()[5])

    to_move = init[f - 1][-q:]

    # for part A, iteratate the reversed list. to_move[::-1]
    for m in to_move:
        init[t - 1] += m

    init[f - 1] = init[f - 1][:-q]

word = ""

for i in init:
    word += i[-1]

print(word)
